### PR TITLE
HBASE-23656 [MERGETOOL] HBASE Support Merge region by pattern

### PR DESCRIPTION
**Pull request by mygood**
Design Objective:
 1. Merge empty region
 2. Neat region
 3. merge expired region
Usage: bin/hbase onlinemerge [--tableName=] [--startRegion=] [--stopRegion=] [--maxRegionSize=] [--maxRegionCreateTime=] [--numMaxMergePlans=] [--targetRegionCount=] [--printExecutionPlan=] [--configMergePauseTime=]
Options:
--h or --h print help
--tableName table name must be not null
--startRegion start region
--stopRegion stop region
--maxRegionSize max region size Unit GB
--maxRegionCreateTime max Region Create Time yyyy/MM/dd HH:mm:ss
--numMaxMergePlans num MaxMerge Plans
--targetRegionCount target Region Count
--configMergePauseTime config Merge Pause Time In milliseconds
--printExecutionPlan Value default is true print execution plans false is execution merge

Examples:
bin/hbase onlinemerge --tableName=test:test1 --startRegion=test:test1,,1576835912332.01d0d6c2b41e204104524d9aec6074fb. --stopRegion=test:test1,bbbbbbbb,1573044786980.0c9b5bd93f3b19eb9bd1a1011ddff66f. --maxRegionSize=0 --maxRegionCreateTime=yyyy/MM/dd HH:mm:ss --numMaxMergePlans=2 --targetRegionCount=4 --printExecutionPlan=false

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 17s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1.4 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 30s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   9m  0s |  branch-1.4 passed  |
| +1 :green_heart: |  compile  |   3m  4s |  branch-1.4 passed with JDK v1.8.0_232  |
| +1 :green_heart: |  compile  |   2m 48s |  branch-1.4 passed with JDK v1.7.0_242  |
| +1 :green_heart: |  checkstyle  |   8m 22s |  branch-1.4 passed  |
| +1 :green_heart: |  shadedjars  |   2m 45s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 35s |  branch-1.4 passed with JDK v1.8.0_232  |
| +1 :green_heart: |  javadoc  |   3m 37s |  branch-1.4 passed with JDK v1.7.0_242  |
| +0 :ok: |  spotbugs  |   2m 29s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  13m  5s |  branch-1.4 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 57s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 43s |  the patch passed with JDK v1.8.0_232  |
| +1 :green_heart: |  javac  |   1m 43s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 46s |  the patch passed with JDK v1.7.0_242  |
| +1 :green_heart: |  javac  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   6m 39s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  2s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   2m 44s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   2m 32s |  Patch does not cause any errors with Hadoop 2.7.7.  |
| +1 :green_heart: |  javadoc  |   2m 34s |  the patch passed with JDK v1.8.0_232  |
| +1 :green_heart: |  javadoc  |   3m 38s |  the patch passed with JDK v1.7.0_242  |
| +1 :green_heart: |  findbugs  |  13m 21s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 208m 46s |  root in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  9s |  The patch does not generate ASF License warnings.  |
|  |   | 300m 47s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1016 |
| Optional Tests | dupname asflicense shellcheck shelldocs javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 709cf1c4a6d2 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1016/out/precommit/personality/provided.sh |
| git revision | branch-1.4 / 879b7ea |
| Default Java | 1.7.0_242 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_232 /usr/lib/jvm/zulu-7-amd64:1.7.0_242 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/testReport/ |
| Max. process+thread count | 4130 (vs. ulimit of 10000) |
| modules | C: hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/console |
| versions | git=1.9.1 maven=3.0.5 shellcheck=0.7.0 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by mygood** [comment]
@saintstack Thank you review code

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 1]
```diff
@@ -0,0 +1,528 @@
+/**
+ * Copyright 2010 The Apache Software Foundation
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: Fix year.

File: bin/hbase.cmd [hunk 2]
```diff
@@ -436,4 +436,5 @@ goto :eof
   echo   mapredcp        Dump CLASSPATH entries required by mapreduce
   echo   version         Print the version
   echo   CLASSNAME       Run the class named CLASSNAME
+  echo   onlinemerge     Run the merge tool
```
Comment (saintstack) [inline bin/hbase.cmd]: What is difference between a merge and an online merge? Why does this have to be here and not in the shell? Shell already has a merge region facility. Add this as an option to the shell command?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+import org.apache.hadoop.hbase.client.RegionLocator;
+import org.apache.hadoop.io.WritableComparator;
+import org.apache.hadoop.util.Tool;
+import org.apache.hadoop.util.ToolRunner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility that can merge any two regions in the same table: adjacent,
+ * overlapping or disjoint. It can also merge every regions, two by two.
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: This is for branch-1? In branch-2, we have a multimerge procedure that can take many (adjacent) regions and squash them up together.

We could have this in branch-1. What would you do for branch-2?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+ */
+@InterfaceAudience.Private
+public class OnlineMergeTool extends Configured implements Tool {
+  static final Logger LOG = LoggerFactory.getLogger(OnlineMergeTool.class);
+  private final int COMPACTPAUSETIME = 180 * 1000;
+  private final int DEFAULTMERGEPAUSETIME = 120 * 1000;
+  private final String COMPACTIONATTRIBUTE = "MAJOR";
+  private final long GB = 1024L * 1024L * 1024L;
+  private final SimpleDateFormat DATE_FORMAT
+      = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: Do ISO8601?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+   */
+  public OnlineMergeTool() throws IOException {
+    this(new HBaseConfiguration());
+  }
+
+  /**
+   * @param conf The current configuration.
+   * @throws IOException If IO problem encountered
+   */
+  public OnlineMergeTool(HBaseConfiguration conf) throws IOException {
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: Yeah, arent' all merges online? Why not just MergeTool?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+    }
+    System.err.println(
+        "Usage: bin/hbase onlinemerge [--tableName=] " +
+            "[--startRegion=] [--stopRegion=] [--maxRegionSize=] " +
+            "[--maxRegionCreateTime=] [--numMaxMergePlans=] " +
+            "[--targetRegionCount=] [--printExecutionPlan=] [--configMergePauseTime=]\n");
+    System.err.println("Options:");
+    System.err.println("--h or --h              print help");
+    System.err.println("--tableName             table name must be not null");
+    System.err.println("--startRegion           start region");
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: Say if name is encoded name or full region name.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+    System.err.println("--stopRegion            stop region");
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/util/OnlineMergeTool.java]: ditto

File: bin/hbase.cmd (hunk 2 above)
Comment (mygood) [reply bin/hbase.cmd]: 当一个大表会有上万的region 为这个表设置一个TTL 会有很多空region 我希望去合并空但是我不希望通过hbase  shell 一个个合并，希望有一个合并的任务去执行。

When a large table has tens of thousands of region to set a TTL for this table, there will be many empty region. I want to merge the empty region, but I don't want to merge them one by one through HBase shell. I want to have a merge task to execute.

Comment (mygood) [reply bin/hbase.cmd]: When a large table has tens of thousands of region to set a TTL for this table, there will be many empty region. I want to merge the empty region, but I don't want to merge them one by one through HBase shell. I want to have a merge task to execute.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 40s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1.4 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 24s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m  1s |  branch-1.4 passed  |
| +1 :green_heart: |  compile  |   1m 44s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 45s |  branch-1.4 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   6m 35s |  branch-1.4 passed  |
| +1 :green_heart: |  shadedjars  |   2m 43s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 35s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   3m 35s |  branch-1.4 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 32s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  13m  2s |  branch-1.4 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 18s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   1m  5s |  root in the patch failed.  |
| -1 :x: |  compile  |   0m 52s |  root in the patch failed with JDK v1.8.0_242.  |
| -1 :x: |  javac  |   0m 52s |  root in the patch failed with JDK v1.8.0_242.  |
| -1 :x: |  compile  |   0m 57s |  root in the patch failed with JDK v1.7.0_252.  |
| -1 :x: |  javac  |   0m 57s |  root in the patch failed with JDK v1.7.0_252.  |
| -1 :x: |  checkstyle  |   1m 30s |  hbase-server: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| -1 :x: |  checkstyle  |   5m  4s |  root: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  shellcheck  |   0m  2s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| -1 :x: |  shadedjars  |   1m 41s |  patch has 114 errors when building our shaded downstream artifacts.  |
| -1 :x: |  hadoopcheck  |   1m 35s |  The patch causes 114 errors with Hadoop v2.7.7.  |
| -1 :x: |  javadoc  |   0m 29s |  hbase-server-jdk1.8.0_242 with JDK v1.8.0_242 generated 10 new + 4 unchanged - 0 fixed = 14 total (was 4)  |
| -1 :x: |  javadoc  |   2m  2s |  root-jdk1.8.0_242 with JDK v1.8.0_242 generated 10 new + 7 unchanged - 0 fixed = 17 total (was 7)  |
| -1 :x: |  javadoc  |   0m 39s |  hbase-server-jdk1.7.0_252 with JDK v1.7.0_252 generated 10 new + 4 unchanged - 0 fixed = 14 total (was 4)  |
| -1 :x: |  javadoc  |   2m 58s |  root-jdk1.7.0_252 with JDK v1.7.0_252 generated 10 new + 10 unchanged - 0 fixed = 20 total (was 10)  |
| -1 :x: |  findbugs  |   0m 33s |  hbase-server in the patch failed.  |
| -1 :x: |  findbugs  |   5m 19s |  root in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  11m 14s |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   |  79m 41s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1016 |
| Optional Tests | dupname asflicense shellcheck shelldocs javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux a07b2c81261b 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1016/out/precommit/personality/provided.sh |
| git revision | branch-1.4 / 8cb278e |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| mvninstall | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-mvninstall-root.txt |
| compile | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-compile-root-jdk1.8.0_242.txt |
| javac | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-compile-root-jdk1.8.0_242.txt |
| compile | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-compile-root-jdk1.7.0_252.txt |
| javac | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-compile-root-jdk1.7.0_252.txt |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-checkstyle-hbase-server.txt |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-checkstyle-root.txt |
| shadedjars | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-shadedjars.txt |
| hadoopcheck | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-javac-2.7.7.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-javadoc-javadoc-hbase-server-jdk1.8.0_242.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-javadoc-javadoc-root-jdk1.8.0_242.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-javadoc-javadoc-hbase-server-jdk1.7.0_252.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/diff-javadoc-javadoc-root-jdk1.7.0_252.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-findbugs-hbase-server.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-findbugs-root.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/testReport/ |
| Max. process+thread count | 230 (vs. ulimit of 10000) |
| modules | C: hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/3/console |
| versions | git=1.9.1 maven=3.0.5 shellcheck=0.7.0 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 38s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1.4 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 23s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m 11s |  branch-1.4 passed  |
| +1 :green_heart: |  compile  |   1m 41s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 46s |  branch-1.4 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   6m 46s |  branch-1.4 passed  |
| +1 :green_heart: |  shadedjars  |   2m 47s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 37s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   3m 38s |  branch-1.4 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 30s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  12m 59s |  branch-1.4 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 18s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 57s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 43s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   1m 43s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 46s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   6m 40s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  1s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   2m 43s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   2m 34s |  Patch does not cause any errors with Hadoop 2.7.7.  |
| +1 :green_heart: |  javadoc  |   2m 35s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   3m 37s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |  13m 25s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 173m 41s |  root in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 14s |  The patch does not generate ASF License warnings.  |
|  |   | 257m 45s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1016 |
| Optional Tests | dupname asflicense shellcheck shelldocs javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 2c331e3ca2f0 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1016/out/precommit/personality/provided.sh |
| git revision | branch-1.4 / 8cb278e |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/4/testReport/ |
| Max. process+thread count | 4241 (vs. ulimit of 10000) |
| modules | C: hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/4/console |
| versions | git=1.9.1 maven=3.0.5 shellcheck=0.7.0 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  11m 50s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  shelldocs  |   0m  0s |  Shelldocs was not available.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1.4 Compile Tests _ |
| +0 :ok: |  mvndep  |   1m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   7m 22s |  branch-1.4 passed  |
| +1 :green_heart: |  compile  |   1m 50s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   1m 53s |  branch-1.4 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   7m 57s |  branch-1.4 passed  |
| +1 :green_heart: |  shadedjars  |   3m  2s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 39s |  branch-1.4 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   3m 50s |  branch-1.4 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 42s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  14m 16s |  branch-1.4 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  7s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 13s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   2m 13s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 12s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   3m 12s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   9m 34s |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  2s |  There were no new shellcheck issues.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   3m 39s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   2m 59s |  Patch does not cause any errors with Hadoop 2.7.7.  |
| +1 :green_heart: |  javadoc  |   2m 43s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   3m 47s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |  15m  8s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 166m 31s |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 59s |  The patch does not generate ASF License warnings.  |
|  |   | 273m 23s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.mapreduce.TestLoadIncrementalHFilesUseSecurityEndPoint |
|   | hadoop.hbase.client.TestAdmin2 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1016 |
| Optional Tests | dupname asflicense shellcheck shelldocs javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux e6f933d0c6ca 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1016/out/precommit/personality/provided.sh |
| git revision | branch-1.4 / 38bf65a |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/testReport/ |
| Max. process+thread count | 4095 (vs. ulimit of 10000) |
| modules | C: hbase-server . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1016/1/console |
| versions | git=1.9.1 maven=3.0.5 shellcheck=0.7.0 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Pull request closed without merging**